tests/integration/components/async-button-test.js

```diff
@@ -0,0 +1,63 @@
+import Ember from 'ember';
+import { moduleForComponent, test } from 'ember-qunit';
+import hbs from 'htmlbars-inline-precompile';
+
+moduleForComponent('async-button', 'Integration | Component | async button', {
+  integration: true
+});
+
+test('it responds to a fulfilled closure promise', function(assert) {
+  let done = assert.async();
+
+  assert.expect(2);
+
+  this.set('default', 'Save');
+  this.set('pending', 'Saving');
+  this.set('fulfilled', 'Saved');
+
+  this.on('closurePromise', function() {
+    return new Ember.RSVP.Promise((fulfill) => {
+      Ember.run.later(() => {
+        fulfill();
+      }, 500);
+    });
+  });
+
+  this.render(hbs`{{async-button default=default pending=pending fulfilled=fulfilled action=(action "closurePromise")}}`);
+  this.$('button').click();
+
+  assert.equal(this.$('button').text().trim(), 'Saving');
+
+  Ember.run.later(() => {
```

Comment:
Awesome, thanks! I'm totally going to refactor my other projects with this, too!